test/unit/azure_rm_test.rb

```diff
@@ -27,5 +29,186 @@ class ForemanAzureRmTest < ActiveSupport::TestCase
+      }
+    }.freeze
+    test "create vm with password and without custom data" do
+      mocked_sdk_methods
+      @test_os_profile.stubs(:custom_data)
+      @test_vm_extension = mock('test_vm_extension')
+      @test_sdk.expects(:create_or_update_vm_extensions).with() do |p1, p2, p3, p4|
+        p1 == "rg1"
+        p4.settings["commandToExecute"] == "$echo testpswd123 | sudo -S echo '\"testuser\" ALL=(ALL) NOPASSWD: ALL' > /etc/sudoers.d/waagent"
+      end.returns(@test_vm_extension)
+      @azure_cr.key_pair = @azure_cr.setup_key_pair
+      server = @azure_cr.create_vm(PSWD_VM_ARGS_WITHOUT_CUSTOM_DATA)
+
+      assert_equal "ervin-golomb", server.name
+      assert_equal "rg1", server.resource_group
+      assert_equal false, server.azure_vm.disable_password_authentication
+      assert_equal 1, server.interfaces.count
+      refute server.azure_vm.os_profile.custom_data.present?
+    end
+
+    PSWD_VM_ARGS_WITH_CUSTOM_DATA_VM_EXT = {
+      "location" => "eastus",
+      "resource_group" => "rg1",
+      "vm_size" => "Standard_A0",
+      "platform" => "Linux",
+      "username" => "testuser",
+      "disable_password_authentication" => "false",
+      "password" => "testpswd123",
+      "name" => "ervin-golomb.example.com",
+      "image_id" => "marketplace://OpenLogic:CentOS:7.5:latest",
+      "user_data" => "echo customData",
+      "script_command" => "sudo sh /var/lib/waagent/custom-script/download/0/myscript.sh",
+      "script_uris" => "https://gist.githubusercontent.com/apuntamb/f4e9ff4e2daf62bc847313b0c64e59f9/raw/73578e1bb7b03237ae1ac6b787b08d00d126adf0/myscript.sh",
+      "interfaces_attributes" => {
+          "0" => {"network" => "[\"/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/first_vnet/subnets/first_subnet\"]", "public_ip" => "Dynamic", "private_ip" => "false"}
+      }
+    }.freeze
+
+    test "create vm with password, custom data and vm extension" do
+      mocked_sdk_methods
+      @test_os_profile.stubs(:custom_data).returns('ZWNobyBjdXN0b21EYXRh')
+      @test_vm_extension = mock('test_vm_extension')
+      @test_sdk.expects(:create_or_update_vm_extensions).with() do |p1, p2, p3, p4|
+        p1 == "rg1"
+        p4.settings["commandToExecute"] == "$echo testpswd123 | sudo -S echo '\"testuser\" ALL=(ALL) NOPASSWD: ALL' > /etc/sudoers.d/waagent ; su - \"testuser\" -c \"sudo sh /var/lib/waagent/custom-script/download/0/myscript.sh\""
+        p4.settings["fileUris"] == ["https://gist.githubusercontent.com/apuntamb/f4e9ff4e2daf62bc847313b0c64e59f9/raw/73578e1bb7b03237ae1ac6b787b08d00d126adf0/myscript.sh"]
+      end.returns(@test_vm_extension)
+      @azure_cr.key_pair = @azure_cr.setup_key_pair
+      server = @azure_cr.create_vm(PSWD_VM_ARGS_WITH_CUSTOM_DATA_VM_EXT)
+      assert_equal "rg1", server.resource_group
+      assert server.azure_vm.os_profile.custom_data.present?
+    end
+
+    SSHKey_VM_ARGS_WITHOUT_CUSTOM_DATA = {
```

Comment:
Can we put this data to the helper too? I think it is related to the mock object that the helper initializes anyway.

Also, it would be awesome if you divide it into blocks, so each test will combine them as needed:

``` ruby
def base_vm_arg
{
# here would be all common properties, like location, resource group and maybe the interface
}
end

def with_password_auth
# here would be the properties needed to enable password auth, like disable_password_authentication and password
end

def with_ssh_key
# here would be ssh related properties, like disable_password_authentication and ssh_key
end
```

the test would be able to combine those:

``` ruby

test 'works with password' do
  vm_args = base_vm_args.merge(with_password_auth)
end

```